File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
class Solution {
public:
    // This returns the node before the N-th node from the end.
    ListNode* findNthNode(ListNode* head, int n ){
        ListNode* lead = head; 
        // Advance lead n steps forward
        for(int i=0;i<n;i++) {
            if (lead) lead = lead->next;
            else return nullptr; // list shorter than n
        }
        ListNode* prev = nullptr;
        ListNode* curr = head;
        // Move both pointers until lead reaches end
        while(lead){
            lead = lead->next;
            prev = curr;
            curr = curr->next;
        }
        // prev now points to node before the one to remove
        return prev;
    }

    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(!head) return nullptr;
        ListNode dummy(0, head); // use dummy for head removal edge case
        ListNode* prev = findNthNode(&dummy, n);
        if(prev && prev->next) {
            ListNode* nodeToRemove = prev->next;
            prev->next = nodeToRemove->next;
            // Optionally: delete nodeToRemove;
        }
        return dummy.next;
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # nth node from end mean. 
        if not head : 
            return None 

        # fast . # -n 
        dummy = ListNode(0, head) 
        slow = fast  = dummy

        while n:
            fast = fast.next 
            n -=1
        
        while fast.next :
            slow = slow.next
            fast = fast.next 

        slow.next = slow.next.next 

        return dummy.next 
        


        